pkg/providers/vault/client.go

```diff
@@ -24,17 +25,39 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+
 	"github.com/edgexfoundry/go-mod-secrets/pkg"
 )
 
+const (
+	AuthTypeHeader = "X-Vault-Token"
+)
+
 // Client defines the behavior for interacting with the Vault REST secret key/value store via HTTP(S).
 type Client struct {
 	HttpConfig SecretConfig
 	HttpCaller Caller
+
+	// internal member variables
+	lc logger.LoggingClient
 }
 
+// package stateful map variable to handle the case of the same caller to have
+// multiple secret clients with potentially the same tokens while renewing token
+// in the background go-routine
+var vaultTokenToCancelFunc = make(map[string]context.CancelFunc)
```

Comment:
Global variables are bad.  Refactor to remove.

Reply:
In this case, we need to keep track of multiple new secret client instances. so we need to use it.

Reply:
You can do this in a way that isn't a global variable.  Remove the global.  Global variables are bad.

Reply:
can you provide some pointers on how this would work?

Reply:
Create a new receiver to hold the global.  Move the NewSecretClient factory function to become a method on the new receiver.  New receiver will need its own factory function.  Call new receiver once during bootstrap; call its NewSecretClient factory method whenever you need a new secret client implementation.

Reply:
Not totally understand what you meant by calling new receiver once during bootstrap.  Is the bootstrap the only place to call this receiver, say  `vaultTokenCancel`.getCancelFunc(), where `vaultTokenCancel` is the new receiver struct ?  Maybe other service like application service will call this new receiver, too?

Reply:
```
type vaultTokenToCancelFuncMap map[string]context.CancelFunc

type secretClientFactory struct {
	vaultTokenToCancelMap vaultTokenToCancelFuncMap
}

func NewSecretClientFactory() *secretClientFactory {
	return &secretClientFactory{
		vaultTokenToCancelMap: make(vaultTokenToCancelFuncMap),
	}
}

func (f *secretClientFactory) NewSecretClient(
	ctx context.Context,
	config SecretConfig,
	lc loggingClient,
	errChan chan<- error) (pkg.SecretClient, error) {

	// existing implementation goes here...

	// if cancel, exists := f.vaultTokenToCancelMap[tokenStr]; exists {
	// 	 cancel()
	// }

	// f.vaultTokenToCancelMap[tokenStr] = cancel
}

// usage 

secretClientFactory := NewSecretClientFactory()
secretClientOne := secretClientFactory.NewSecretClient(...)
secretClientTwo := secretClientFactory.NewSecretClient(...)
```